### PR TITLE
openstack: wait for volume to be removed

### PR DESCRIPTION
**Pull request by bennyz**
Before removing the snapshot attempt to make sure the volume has been removed

### CODE REVIEW

File: pkg/controller/plan/adapter/openstack/client.go [hunk 1]
```diff
@@ -536,6 +533,51 @@ func (r *Client) cleanup(vm *libclient.VM, originalVolumeID string) (err error)
 		err = nil
 		return
 	}
+
+	// Now we need to wait for the volume to be removed
+	condition := func() (done bool, err error) {
+		volume, err := r.getVolumeFromSnapshot(vm, snapshot.ID)
+		if err != nil {
+			if errors.Is(err, ResourceNotFoundError) {
+				r.Log.Info("volume not found, we are done")
```
Comment (ahadas) [inline pkg/controller/plan/adapter/openstack/client.go:542]: this log can be confusing, need to rephrase it

File: pkg/controller/plan/adapter/openstack/client.go [hunk 2]
```diff
@@ -536,6 +533,51 @@ func (r *Client) cleanup(vm *libclient.VM, originalVolumeID string) (err error)
...
+		default:
+			err = UnexpectedVolumeStatusError
+			r.Log.Error(err, "checking the volume",
+				"vm", vm.Name, "volume", volume.Name, "status", volume.Status)
+			return
+		}
+		return
+	}
+
+	// TODO add config
```
Comment (ahadas) [inline pkg/controller/plan/adapter/openstack/client.go:565]: yep

File: pkg/controller/plan/adapter/openstack/client.go [hunk 3]
```diff
@@ -536,6 +533,51 @@ func (r *Client) cleanup(vm *libclient.VM, originalVolumeID string) (err error)
...
+		Steps:    100,
+	}
+
+	err = wait.ExponentialBackoff(backoff, condition)
+	if err != nil {
+		err = liberr.Wrap(err)
+		r.Log.Error(err, "waiting for the volume to be removed",
+			"vm", vm.Name, "snapshotID", snapshot.ID)
+		err = nil
+		return
```
Comment (ahadas) [inline pkg/controller/plan/adapter/openstack/client.go:578]: and what happens at this point?

**Comment by ahadas** [comment]
@bennyz I've noticed we had another issue with the snapshot status [here](https://github.com/kubev2v/forklift/actions/runs/6682009814/job/18156703604), could you please take a look?

**Comment by bennyz** [comment]
> @bennyz I've noticed we had another issue with the snapshot status [here](https://github.com/kubev2v/forklift/actions/runs/6682009814/job/18156703604), could you please take a look?

looks like the usual env issues

**Comment by ahadas** [comment]
> > @bennyz I've noticed we had another issue with the snapshot status [here](https://github.com/kubev2v/forklift/actions/runs/6682009814/job/18156703604), could you please take a look?
> 
> looks like the usual env issues

I didn't see it for a while.. and what was the problem in that case with the snapshot?..

**Comment by bennyz** [comment]
> > > @bennyz I've noticed we had another issue with the snapshot status [here](https://github.com/kubev2v/forklift/actions/runs/6682009814/job/18156703604), could you please take a look?
> > 
> > 
> > looks like the usual env issues
> 
> I didn't see it for a while.. and what was the problem in that case with the snapshot?..

I only see stuff like
```
2023-10-29 08:06:34.833 1313 ERROR oslo.messaging._drivers.impl_rabbit [-] [13679efd-cea0-4a1e-b931-4d48e7ac28e4] AMQP server on 127.0.0.1:5672 is unreachable: Too many heartbeats missed. Trying again in 1 seconds.: amqp.exceptions.ConnectionForced: Too many heartbeats missed
```

File: pkg/controller/plan/adapter/openstack/client.go [hunk 4]
```diff
@@ -272,12 +274,13 @@ func (r *Client) PreTransferActions(vmRef ref.Ref) (ready bool, err error) {
...
-				r.Log.Error(err, "cleaning the image",
-					"vm", vm.Name, "image", image.Name)
-				return
-			}
+			go func() {
+				err := r.cleanup(vm, originalVolumeID)
+				if err != nil {
+					r.Log.Error(err, "failed to cleanup")
+				}
+			}()
```
Comment (ahadas) [inline pkg/controller/plan/adapter/openstack/client.go:282]: @bennyz can you please remind me why we moved this into a goroutine? is it an optimization that is meant to reduce the execution time?

File: pkg/controller/plan/adapter/openstack/client.go (hunk 1 above)
Comment (ahadas) [reply pkg/controller/plan/adapter/openstack/client.go:558]:
```suggestion
				r.Log.Info("volume doesn't exist, assuming it was removed")
```

File: pkg/controller/plan/adapter/openstack/client.go (hunk 4 above)
Comment (bennyz) [reply pkg/controller/plan/adapter/openstack/client.go:282]: mainly because it's a best effort operation which we don't need to wait for

Comment (ahadas) [reply pkg/controller/plan/adapter/openstack/client.go:283]: ah is it a blocking operation?

Comment (bennyz) [reply pkg/controller/plan/adapter/openstack/client.go:283]: yes, specifically the waiting part

Comment (ahadas) [reply pkg/controller/plan/adapter/openstack/client.go:283]: hehe, oops, I forgot about the rest of this PR as I'm concentrated on #491 ..

**Review by ahadas — APPROVED** [review]

**Pull request merged**